src/V3Dfg.h

```diff
@@ -0,0 +1,698 @@
+    // the name of the graph which is included in the output.
+    void dumpDotFile(const string& fileName, const string& label = "") const;
+    // Dump graph in Graphviz format into a new automatically numbered debug file. 'label' is
+    // added to the name of the graph, which is included in the file name and the output.
+    void dumpDotFilePrefixed(const string& label = "") const;
+    // Dump upstream (source) logic cone starting from given vertex into a file with the given
+    // 'fileName'. 'name' is the name of the graph, which is included in the output.
+    void dumpDotUpstreamCone(const string& fileName, const DfgVertex& vtx,
+                             const string& name = "") const;
+    // Dump all individual logic cones driving external variables in Graphviz format into separate
+    // new automatically numbered debug files. 'label' is added to the name of the graph, which is
+    // included in the file names and the output. This is useful for very large graphs that are
+    // otherwise difficult to browse visually due to their size.
+    void dumpDotAllVarConesPrefixed(const string& label = "") const;
+};
+
+//------------------------------------------------------------------------------
+// Dataflow graph edge
+//------------------------------------------------------------------------------
+
+class DfgEdge final {
+    friend class DfgVertex;
+    template <size_t Arity>
+    friend class DfgVertexWithArity;
+
+    DfgEdge* m_nextp = nullptr;  // Next edge in sink list
+    DfgEdge* m_prevp = nullptr;  // Previous edge in sink list
+    DfgVertex* m_sourcep = nullptr;  // The source vertex driving this edge
+    DfgVertex* const m_sinkp;  // The sink vertex. The sink owns the edge, so immutable
+
+    explicit DfgEdge(DfgVertex* sinkp)  // The sink vertices own the edges, hence private
+        : m_sinkp{sinkp} {}
+
+public:
+    // The source (driver) of this edge
+    DfgVertex* sourcep() const { return m_sourcep; }
+
+    // The sink (consumer) of this edge
+    DfgVertex* sinkp() const { return m_sinkp; }
+
+    // Remove driver of this edge
+    void unlinkSource();
+
+    // Relink this edge to be driven from the given new source vertex
+    void relinkSource(DfgVertex* newSourcep);
+};
+
+//------------------------------------------------------------------------------
+// Dataflow graph vertex
+//------------------------------------------------------------------------------
+
+// Reuse the generated type constants
+using DfgType = VNType;
+
+// Base data flow graph vertex
+class DfgVertex VL_NOT_FINAL {
+    friend class DfgGraph;
+    friend class DfgEdge;
+    friend class DfgVVisitor;
+
+    // STATE
+    V3ListEnt<DfgVertex*> m_verticesEnt;  // V3List handle of this vertex, kept under the DfgGraph
+protected:
+    DfgEdge* m_sinksp = nullptr;  // List of sinks of this vertex
+    FileLine* const m_filelinep;  // Source location
+    uint32_t m_width = 0;  // Width of the result of this vertex
```

Comment:
I could, but:
- I need to pick some canonical packed dtype to represent fixed width types (for example, `logic [31:0]`, `logic [3:0][7:0]`, and `int unsigned` - and even their signed versions - are all the same as far as the DFG IR is concerned, but I need these to be the same otherwise it's intractable to get it right. If you are ok we just use the same packed type for all then this is doable.
- this would mean a back reference to the type table so that needs to be kept live. Currently this is a non issue as we don't collect dead types while a DfgGraph is live.

If you are ok with this I can change  the width to a dtype pointer (but keep a constructor that take a width as well for convenience where appropriate), then add a `width()` method that crashes if the type is not the canonical packed type (I don't want references to unpacked arrays to have a width for example)

Reply:
I'm ok we use the same packed type for all compatible types.

Reply:
I moved to dtypes. Messes with my parallelisation plan due to the interned types in the type table, but I will sort that out another time.